gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/VectoredIOImpl.java

```diff
@@ -0,0 +1,371 @@
+public class VectoredIOImpl implements Closeable {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final BlockingQueue taskQueue = new LinkedBlockingQueue<Runnable>();
+  private final VectoredReadOptions vectoredReadOptions;
+  private final GhfsGlobalStorageStatistics storageStatistics;
+  private final FileSystem.Statistics statistics;
+  private ExecutorService boundedThreadPool;
+
+  public VectoredIOImpl(
+      VectoredReadOptions vectoredReadOptions,
+      GhfsGlobalStorageStatistics storageStatistics,
+      FileSystem.Statistics statistics) {
+    this.vectoredReadOptions = vectoredReadOptions;
+    // same fixedThreadPool executor, but provided a way to query task queue
+    this.boundedThreadPool =
+        new ThreadPoolExecutor(
+            vectoredReadOptions.getReadThreads(),
+            vectoredReadOptions.getReadThreads(),
+            0L,
+            TimeUnit.MILLISECONDS,
+            taskQueue,
+            new ThreadFactoryBuilder().setNameFormat("gcs-async-vectored-pool-%d").build());
+    this.statistics = statistics;
+    this.storageStatistics = storageStatistics;
+  }
+
+  /**
+   * Reads data from Google Cloud Storage using vectored I/O operations.
+   *
+   * @param ranges List of file ranges to read.
+   * @param allocate Function to allocate ByteBuffer for reading.
+   * @throws IOException if an I/O error occurs.
+   */
+  /**
+   * Reads data from Google Cloud Storage using vectored I/O operations.
+   *
+   * @param ranges List of file ranges to read.
+   * @param allocate Function to allocate ByteBuffer for reading.
+   * @param gcsFs GCFS implementation to use while creating channel and reading content for ranges.
+   * @param fileInfo FileInfo of the gcs object agaisnt which range request are fired, this can be
+   *     null for some code path fall back to URI path provided.
+   * @param gcsPath URI of the gcs object for which the range requests are fired.
+   * @throws IOException If invalid range is requested, offset<0.
+   */
+  public void readVectored(
+      List<? extends FileRange> ranges,
+      IntFunction<ByteBuffer> allocate,
+      GoogleCloudStorageFileSystem gcsFs,
+      FileInfo fileInfo,
+      @Nonnull URI gcsPath)
+      throws IOException {
+    ReadChannelProvider channelProvider = new ReadChannelProvider(gcsFs, fileInfo, gcsPath);
+    readVectored(ranges, allocate, channelProvider);
+  }
+
+  private void readVectored(
+      List<? extends FileRange> ranges,
+      IntFunction<ByteBuffer> allocate,
+      ReadChannelProvider channelProvider)
+      throws IOException {
+    List<? extends FileRange> sortedRanges = validateNonOverlappingAndReturnSortedRanges(ranges);
+    for (FileRange range : ranges) {
+      // TODO(user): upgrade to use validateAndSortRanges
+      validateRangeRequest(range);
+      storageStatistics.updateStats(
+          GhfsStatistic.STREAM_READ_VECTORED_READ_RANGE_SIZE,
+          range.getLength(),
```

Comment:
No.

> This will always be same within an executor, correct?

why one would assume that?

>  what is the value of having a duration metric for this?

It's not a duration but a gauge metric, value addition in getting the (min,max,mean) stat is that we would get to know the various sizes of range being requested and try optimizing the connector's performance for the same.

e.g. let's take two scenarios 
1. avg range size: 100MB
2. avg rang size: 100KB

in scenario (1) It might make sense to further divide the ranges at client side to get better performance and we will be able to make enhancement to our parquet-mr code to do the same

in scenario (2) It makes sense to increase the minSeek value to even higher to get better performance because majority of the time is getting invested in opening the channel and actual read cost(in terms of time) is lower.